<a href="https://colab.research.google.com/github/spasatel13/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module2-make-features/Philipp_Libenson_DS21_W1_D2_Make_Features_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 2*

---

# Learning Objectives

- Student should be able to understand the purpose of feature engineering
- Student should be able to work with strings in pandas
- Student should be able to work with dates and times in pandas
- Student should be able to modify or create columns of a dataframe using the `.apply()` function


Helpful Links:
- [Minimally Sufficient Pandas](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)
- [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)
- Python Data Science Handbook
  - [Chapter 3.10](https://jakevdp.github.io/PythonDataScienceHandbook/03.10-working-with-strings.html), Vectorized String Operations
  - [Chapter 3.11](https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html), Working with Time Series
- [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)

# Let's continue to explore feature engineering with the Ames Housing data.  Import the dataset and name it "house".

###1) Import the Ames Housing Data here: https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv

In [4]:
#Code here
import pandas as pd
house=pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')

In [ ]:
!curl https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv

In [7]:
house.shape

(1460, 81)

###2) Print the first five rows of the house dataset.

In [8]:
#Code here
house.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

###3) Create a new variable called Large_Lot that takes on the value 1 when LotArea is gerater than or equal to 10,000 and the value 0 when LotArea is less than 10,000.

In [15]:
#Code here
condition=house['LotArea']>=10000
house['Large_Lot']= [1 if area>=10000 else 0 for area in house['LotArea']]
house.head()

Id  MSSubClass MSZoning  ...  SaleCondition  SalePrice Large_Lot
0   1          60       RL  ...         Normal     208500         0
1   2          20       RL  ...         Normal     181500         0
2   3          60       RL  ...         Normal     223500         1
3   4          70       RL  ...        Abnorml     140000         0
4   5          60       RL  ...         Normal     250000         1

[5 rows x 82 columns]

###4) Create a new variable called Total_Bathrooms that contains the total number of full and half bathrooms in the house.

In [18]:
#Code here
house['Total_Bathrooms']=house['FullBath']+house['HalfBath']
house.head()

Id  MSSubClass MSZoning  ...  SalePrice  Large_Lot Total_Bathrooms
0   1          60       RL  ...     208500          0               3
1   2          20       RL  ...     181500          0               2
2   3          60       RL  ...     223500          1               3
3   4          70       RL  ...     140000          0               1
4   5          60       RL  ...     250000          1               3

[5 rows x 83 columns]

###5) Create a new variable called Summer_Sale that equals 1 if the sale month was in June, July or August ( MoSold = 6, 7 or 8) and 0 if the sale month was in September - May.

In [21]:
#code here
house["Summer_Sale"]=[1 if month>=6 and month<=8 else 0 for month in house['MoSold']]
house.head(10)

Id  MSSubClass MSZoning  ...  Large_Lot  Total_Bathrooms Summer_Sale
0   1          60       RL  ...          0                3           0
1   2          20       RL  ...          0                2           0
2   3          60       RL  ...          1                3           0
3   4          70       RL  ...          0                1           0
4   5          60       RL  ...          1                3           0
5   6          50       RL  ...          1                2           0
6   7          20       RL  ...          1                2           1
7   8          60       RL  ...          1                3           0
8   9          50       RM  ...          0                2           0
9  10         190       RL  ...          0                1           0

[10 rows x 84 columns]

#Now we'll revisit the LendingClub data.

Remember to import the data by first unzipping the file:

In [22]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-10-05 04:29:17--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 52.27.223.142, 54.68.156.83, 34.211.109.164
Connecting to resources.lendingclub.com (resources.lendingclub.com)|52.27.223.142|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [             <=>    ]  22.30M  2.17MB/s    in 10s     

2020-10-05 04:29:28 (2.14 MB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [23387713]



In [23]:
!unzip LoanStats_2018Q4.csv.zip


Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


In [35]:
# Read in the CSV
import pandas as pd
df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1, skipfooter=2, engine='python')

print(df.shape)
df.head()

(128378, 144)


id  member_id  ...  settlement_percentage  settlement_term
0 NaN        NaN  ...                    NaN              NaN
1 NaN        NaN  ...                    NaN              NaN
2 NaN        NaN  ...                    NaN              NaN
3 NaN        NaN  ...                    NaN              NaN
4 NaN        NaN  ...                    NaN              NaN

[5 rows x 144 columns]

#Approaching a problem two different ways.

In the guided project, we learned how to find the earliest credit year using the built-in pandas date-time format.  However, there is often more than one way to come up with the same solution.  







In the following questions we will work through the steps to create a new variables called Earliest_Credit_Year and issue_year where we use the .split() function to remove the month and report back just the year as a float data type.  Then we'll use those variables to calculate the length of credit history in years.  

If we do everything correctly, we should calculate *about* the same length of longest credit history that we did in class working with the date-time format.  The answers won't exactly match because we are making different assumptions about when in each month and year the loans were taken out, but it will give you a flavor for different ways of approaching the same problem.

###6) Create a simple test case where credit = 'Jun-1979'.  

Use the .split('-') function to separate the month and year parts of the data and print the year.

Hint: Name the results of the .split('-') function "fields".  fields[0] will return the month part of the date and fields[1] will return the year part of the date.

In [26]:
#code here
fields='Jun-1979'.split('-')
fields

['Jun', '1979']

###7) Use your answer in 6) to write a function called credit_yr that takes in the contents of a cell formatted 'month-year' and returns the year.

Run your function using the simple test case in 6).

In [31]:
#Code here
def credit_yr(date):
  return date.split('-')[1]
x=credit_yr('Jun-1979')
print(x)

1979


###8) Use the .apply() function to apply the function to every cell in the earliest_cr_line variable.

In [48]:
#code here
df['Earliest_Credit_Year']=df['earliest_cr_line'].apply(credit_yr)
df.head()

id  member_id  ...  settlement_term  Earliest_Credit_Year
0 NaN        NaN  ...              NaN                  2007
1 NaN        NaN  ...              NaN                  2002
2 NaN        NaN  ...              NaN                  1999
3 NaN        NaN  ...              NaN                  2011
4 NaN        NaN  ...              NaN                  1993

[5 rows x 145 columns]

###9) Print the top 5 rows of the Earliest_Credit_Year and earliest_cr_line variables to make sure the variables were created correctly.

In [50]:
#code here
df[['earliest_cr_line','Earliest_Credit_Year']].head()

earliest_cr_line Earliest_Credit_Year
0         Oct-2007                 2007
1         Jan-2002                 2002
2         Feb-1999                 1999
3         Aug-2011                 2011
4         Oct-1993                 1993

### 10) Use your function to create a new variable called Issue_Year from the issue_d variable.

In [51]:
#code here
df['issue_Year']=df['issue_d'].apply(credit_yr)
df.head()

id  member_id  loan_amnt  ...  settlement_term  Earliest_Credit_Year issue_Year
0 NaN        NaN       5000  ...              NaN                  2007       2018
1 NaN        NaN      10000  ...              NaN                  2002       2018
2 NaN        NaN      20000  ...              NaN                  1999       2018
3 NaN        NaN       6500  ...              NaN                  2011       2018
4 NaN        NaN      25000  ...              NaN                  1993       2018

[5 rows x 146 columns]

###11) Caluculate length of credit history in years and in days.

Hint: If you are getting an error message, make sure to check the data type of the years output by your functions.

In [59]:
#Code here
   # Two Ways to convert String series to Int
df['length_of_credit_history_years']=(df["issue_Year"]).astype(int)- pd.to_numeric(df['Earliest_Credit_Year'])
df['length_of_credit_history_days']=df['length_of_credit_history_years']*365.25
df.head()

id  member_id  ...  length_of_credit_history_years  length_of_credit_history_days
0 NaN        NaN  ...                              11                        4017.75
1 NaN        NaN  ...                              16                        5844.00
2 NaN        NaN  ...                              19                        6939.75
3 NaN        NaN  ...                               7                        2556.75
4 NaN        NaN  ...                              25                        9131.25

[5 rows x 149 columns]

###12) Calculate the maximum credit length and compare it to we got in the guided project.  

Note - it won't be *exactly* the same because both methods are working in different ways, but they will give you a "pretty close" answer.

In [61]:
#code here
df[['length_of_credit_history_years','length_of_credit_history_days']].max()

length_of_credit_history_years       68.0
length_of_credit_history_days     24837.0
dtype: float64

Guided Project Values: 25,171 days, 68.96 years of credit history.\
Assignment Project Values 24,837 days, 68 years of credit history.\
Pretty close. 

#Portfolio Project Assignment

Start looking for interesting datasets!  

Pretty much every major US governmental agency (CDC, Bureau of Labor Statistics, The Census, etc.) maintains a treasure trove of data.

There's also:

Datasets from the University of Florida Department of Statistics: http://users.stat.ufl.edu/~winner/datasets.html

Pew Research: https://www.pewresearch.org/

General Social Survey: https://gss.norc.org/

Ancestry.com Free data collections: https://www.ancestry.com/search/categories/freeindexacom/?clickref=1101lbhdWxys&key=Uhttps%3A%2F%2Fwww.ancestry.com%2Fsearch%2Fcategories%2Ffreeindexacom%2F%3Fclickref%3D1101lbhdWxys&o_lid=01011l4pVw&o_sch=Affiliate+External&o_xid=01011l4pVw

Popular baby names by year since the founding of the SSA: https://www.babynamewizard.com/ 


https://data.world/

https://www.kaggle.com/datasets

https://catalog.data.gov/dataset

https://datasetsearch.research.google.com/

https://archive.ics.uci.edu/ml/index.php

https://github.com/plotly/datasets

https://lionbridge.ai/datasets/20-best-image-datasets-for-computer-vision/

https://www.visualdata.io/

https://pathmind.com/wiki/open-datasets

https://blog.playment.io/top-open-image-dataset/

https://blog.cambridgespark.com/

https://www.analyticsvidhya.com/blog/2018/03/

https://data-flair.training/blogs/machine-learning-datasets/





